### ISSUE TITLE
ProteinQuantifier and protein group probability

### ISSUE THREAD

**Issue by zhuchcn** [OpenMS/OpenMS#6038]
We recently are trying to reproduce the results of a published study with TMT-10. We used MSFraggerAdapter, followed by Percolator to calculate PEP, and ProteinProphet to do protein grouping. If we take the protXML from ProteinProphet, convert it to idXML and use directly in ProteinQuantifier with `-protein_groups`, the protein quantitation result has most proteins being only quantified in one plex, and only a few are quantified in all TMT plexes. See the barplot below:

![image](https://user-images.githubusercontent.com/29579625/163100104-d67836a4-453e-4277-8ee6-96cef0af3ca3.png)

After looking into PeptideProphet, we found that for a protein group, it usually assign only one protein, but other proteins of the sample groups are also written into the protXML file with the probability score as zero (that they called the "unneeded" proteins). When the protXML is converted to idXML, those unneeded proteins are all kept as `protein_group` and `indistinguishable_protein`.  For example, let's say protein `ENSP00000308897.7` and `ENSP00000329630.4` indistinguishable proteins, in the idXML file, they both are present, like this below. `PH_0` is `ENSP00000308897.7` and `PH_` is `ENSP00000329630.4`. So I think this causes ProteinQuantifier not being able to use any shared peptides of these two proteins to quantify `ENSP00000308897.7`.

```xml
<UserParam type="string" name="protein_group_0" value="0.99998,PH_0"/>
<UserParam type="string" name="protein_group_1" value="0,PH_1"/>
```

We then added an filter step on the ProteinProphet idXML file to remove the protein groups with the probability score being zero and then use the filtered idXML as protein groups, the results look much better, as most proteins a quantified in most of the plexes. We also tried Fido and saw very similar pattern, that if we use the original protein grouping idXML, most proteins are quantified in only one plex, and only when we filter out protein groups with low probability score, it starts to look decent. But the problem for Fido is a little harder, because it doesn't seem to specifically assign "unneeded" proteins to zero, so after some test runs, I was able to find the best cutoff being 0.6. See figure below.

![image](https://user-images.githubusercontent.com/29579625/163101683-0317ef48-f2d2-48fb-acfa-21211bd4f7b9.png)

For our pipeline, our workaround is just a python script to remove protein groups with a given cutoff. But it would be great if ProteinQuantifier can support a filter of protein group probability already. Otherwise anyone using this tool will probably have the same issue.

**Comment by zhuchcn** [comment]
> ProteinQuantifier will only use peptides that are unique or unique to an indistinguishable group (e.g. only shared by all the proteins in such a group). In the second case, it will report a quantity for the whole group.

This makes sense to me, but I think the problem is a little different. Consider the example I gave above of `PH_0` and `PH_1`. They actually belong to the same protein group, but just ProteinProphet reports both, as `PH_0` being the leader protein with probability of 0.99998, and `PH_1` being the leader protein with probability of 0 (unneeded). So in ProteinQuantifier, for the peptide shared by `PH_0` and `PH_`, when it finds the leader protein, the leader of `PH_0` is `PH_0`, but that of `PH_1` becomes `PH_1`, then there is no unique leader protein, so ProteinQuantifier thinks they belong to different protein groups. Let me know if I'm not interpreting it correctly!

> Maybe the pepXML parser needs to do a (better?) distinction between indistinguishable and more general groups.

Maybe the protXML parser as well? Right now indistinguishable proteins don't have any probability scores at all when converting protXML to idXML. But general protein groups do have probability.

**Comment by jpfeuffer** [comment]
Yes I forgot that they have the additional protXML. I mean that.

Only if the protXML converter outputs
<UserParam type="string" name="indistinguishable _protein_group_0" value="0.99998,PH_0,PH_1"/>
their peptides will be used for quantification of the whole group.
Every other scenario will lead to their shared peptides being discarded.
The problem is that (AFAIR) protXML does not annotate the type of groups. So we would probably need to run our internal grouping function to distinguish between different group types.

**Comment by zhuchcn** [comment]
Yeah, seems like ProteinProphet gives something like this:

```xml
<UserParam type="string" name="indistinguishable _protein_group_0" value="0.99998,PH_0"/>
<UserParam type="string" name="indistinguishable _protein_group_0" value="0,PH_1"/>
```

That's why filtering the ones with zero works for me. Btw, I just tried with `-greedy_group_resolution`, it helps a little but still not great. There are still a lot of proteins only quantified in one plex. Results like this is just hard for downstream analysis.

![image](https://user-images.githubusercontent.com/29579625/163252796-615855c3-d208-459e-acce-d33ea593d9b6.png)

**Comment by jpfeuffer** [comment]
I can look at the protXML parsing but now that you are mentioning "missing in one plex" again (which I kind of ignored for now), I don't think this has to do anything with grouping or inference.
This is caused by missing channels and you probably need to reevaluate your IsobaricAnalyzer settings. There should be a summary at the end of the tool's log about the number of quantified channels.

**Comment by zhuchcn** [comment]
I found this for one fraction of one plex. Doesn't look too bad? I checked other plexes and fractions and some are a little worse, but they are all above 92%. 

```
IsobaricQuantifier: skipped 409 of 37477 selected scans due to lack of reporter information:
IsobaricQuantifier: channels with signal
  ch 126 : 36809 / 37477 (98%)
  ch 127N: 36779 / 37477 (98%)
  ch 127C: 36831 / 37477 (98%)
  ch 128N: 36831 / 37477 (98%)
  ch 128C: 36842 / 37477 (98%)
  ch 129N: 36857 / 37477 (98%)
  ch 129C: 36885 / 37477 (98%)
  ch 130N: 36839 / 37477 (98%)
  ch 130C: 36782 / 37477 (98%)
  ch 131 : 36813 / 37477 (98%)
```

The motivation that got me thinking of this is an issue of protein grouping and quantitation is the debug log output by ProteinQuantifier (with `-debug`), with a lot of peptides with "LEADERS EMPTY". Example below:

```
PeptideAndProteinQuant.cpp(358): LEADERS EMPTY:
PeptideAndProteinQuant.cpp(361): ENSP00000266544.5|ENST00000266544.10|ENSG00000139180.11|OTTHUMG00000168224.7|OTTHUMT00000398900.4|NDUFA9-201|NDUFA9|377
PeptideAndProteinQuant.cpp(361): ENSP00000497305.1|ENST00000648836.1|ENSG00000255639.3|OTTHUMG00000168352.4|OTTHUMT00000500926.1|AC005833.1-201|AC005833.1|567
```

In the ProteinProphet output (protXML), I see these record below. `ENSP00000266544.5` is the leader protein, and other proteins including `ENSP00000497305.1` have shared peptides with it.

```xml
<protein_group group_number="2715" probability="1.0000">
    <protein protein_name="ENSP00000266544.5|ENST00000266544.10|ENSG00000139180.11|OTTHUMG00000168224.7|OTTHUMT00000398900.4|NDUFA9-201|NDUFA9|377" n_indistinguishable_proteins="1" probability="1.0000" percent_coverage="78.2" unique_stripped_peptides="AAAAQSR+AAAAQSRVVR+AAAAQSRVVRVLSMSR+AIEVLRR+DAFPEAIIVKPSDIFGR+DAFPEAIIVKPSDIFGREDR+DKVERMHITDMK+FGPIPLGSLGWK+FLNSFASMHR+FLNSFASMHRFGPIPLGSLGWK+GGRSSVSGIVATVFGATGFLGR+GIVNAVK+GIVNAVKDPDANGK+IPQAIAQLSK+LPHLPGLEDLGIQATPLELK+MAAAAQSRVVR+MAAAAQSRVVRVLSMSR+MGSQVIIPYR+MHITDMK+NFDFEDVFVK+QLHHALMPHGK+QPVYVVDVSK+QPVYVVDVSKGIVNAVK+SAITAIATSVCHGPPCR+SFAFVGPSR+SSVSGIVATVFGATGFLGR+VFEISPFEPWITR+VVQHSNVVINLIGR+VVQHSNVVINLIGRDWETK+WLSAEIEDVKPAK+YDIMHLRPMGDLGQLLFLEWDAR+YIFAVAHR+YVVNHLGR" group_sibling_id="a" total_number_peptides="161" total_number_distinct_peptides="161" pct_spectrum_ids="0.011" confidence="1.000">
         <parameter name="prot_length" value="377"/>
         <annotation protein_description="Protein No. 1"/>
         ...
         <peptide peptide_sequence="FGPIPLGSLGWK" charge="0" initial_probability="0.9990" nsp_adjusted_probability="0.9992" fpkm_adjusted_probability="0.9992" weight="1.00" group_weight="1.00" is_nondegenerate_evidence="N" n_enzymatic_termini="2" n_sibling_peptides="21.42" n_sibling_peptides_bin="8" n_instances="10" exp_tot_instances="9.22" is_contributing_evidence="Y">
            <peptide_parent_protein protein_name="ENSP00000497305.1|ENST00000648836.1|ENSG00000255639.3|OTTHUMG00000168352.4|OTTHUMT00000500926.1|AC005833.1-201|AC005833.1|567"/>
             <indistinguishable_peptide peptide_sequence="FGPIPLGSLGWK" charge="2" calc_neutral_pep_mass="1729.03">
             <modification_info modified_peptide="n[230]FGPIPLGSLGWK[357]"/>
             </indistinguishable_peptide>
             <indistinguishable_peptide peptide_sequence="FGPIPLGSLGWK" charge="3" calc_neutral_pep_mass="1729.03">
             <modification_info modified_peptide="n[230]FGPIPLGSLGWK[357]"/>
             </indistinguishable_peptide>
         </peptide>
         ...
    </protein>
</protein_group>
```

But I also saw this record below, so `ENSP00000497305.1` being the leader protein, while `ENSP00000266544.5` is the subsuming protein, but jus the probability of this is zero.

```xml
<protein_group group_number="63607" probability="0.0000">
    <protein protein_name="ENSP00000497305.1|ENST00000648836.1|ENSG00000255639.3|OTTHUMG00000168352.4|OTTHUMT00000500926.1|AC005833.1-201|AC005833.1|567" n_indistinguishable_proteins="1" probability="0.0000" unique_stripped_peptides="AAAAQSR+AAAAQSRVVR+AAAAQSRVVRVLSMSR+DAFPEAIIVKPSDIFGR+DAFPEAIIVKPSDIFGREDR+FGPIPLGSLGWK+FLNSFASMHR+FLNSFASMHRFGPIPLGSLGWK+GGRSSVSGIVATVFGATGFLGR+GIVNAVK+GIVNAVKDPDANGK+IPQAIAQLSK+MAAAAQSRVVR+MAAAAQSRVVRVLSMSR+MGSQVIIPYR+NFDFEDVFVK+QLHHALMPHGK+QPVYVVDVSK+QPVYVVDVSKGIVNAVK+SAITAIATSVCHGPPCR+SFAFVGPSR+SSVSGIVATVFGATGFLGR+VFEISPFEPWITR+VVQHSNVVINLIGR+VVQHSNVVINLIGRDWETK+YDIMHLRPMGDLGQLLFLEWDAR+YIFAVAHR+YVVNHLGR" group_sibling_id="a" total_number_peptides="0" total_number_distinct_peptides="0" subsuming_protein_entry="ENSP00000266544.5|ENST00000266544.10|ENSG00000139180.11|OTTHUMG00000168224.7|OTTHUMT00000398900.4|NDUFA9-201|NDUFA9|377" confidence="1.0000">
        <parameter name="prot_length" value="567"/>
        <annotation protein_description="ENSP00000497305.1 ENST00000648836.1 ENSG00000255639.3 OTTHUMG00000168352.4 OTTHUMT00000500926.1 AC005833.1-201 AC005833.1 567"/>
        <peptide peptide_sequence="FGPIPLGSLGWK" charge="0" initial_probability="0.9990" nsp_adjusted_probability="0.9928" fpkm_adjusted_probability="0.9928" weight="0.00" group_weight="0.00" is_nondegenerate_evidence="N" n_enzymatic_termini="2" n_sibling_peptides="0.00" n_sibling_peptides_bin="0" n_instances="10" exp_tot_instances="9.22" is_contributing_evidence="N">
        <peptide_parent_protein protein_name="ENSP00000266544.5|ENST00000266544.10|ENSG00000139180.11|OTTHUMG00000168224.7|OTTHUMT00000398900.4|NDUFA9-201|NDUFA9|377"/>
            <indistinguishable_peptide peptide_sequence="FGPIPLGSLGWK" charge="2" calc_neutral_pep_mass="1729.03">
            <modification_info modified_peptide="n[230]FGPIPLGSLGWK[357]"/>
            </indistinguishable_peptide>
            <indistinguishable_peptide peptide_sequence="FGPIPLGSLGWK" charge="3" calc_neutral_pep_mass="1729.03">
            <modification_info modified_peptide="n[230]FGPIPLGSLGWK[357]"/>
            </indistinguishable_peptide>
        </peptide>
        ...
    </protein>
</protein_group>
```

So when the protXML file got converted to idXML, the both records are kept, ending up with this (only relevant information kept):

```xml
<ProteinIdentification score_type="ProteinProphet probability" higher_score_better="true" significance_threshold="0" >
    ...
    <ProteinHit id="PH_9008" accession="ENSP00000266544.5|ENST00000266544.10|ENSG00000139180.11|OTTHUMG00000168224.7|OTTHUMT00000398900.4|NDUFA9-201|NDUFA9|377" score="1.0" coverage="78.200000000000003" sequence="" >
    </ProteinHit>
    <ProteinHit id="PH_186555" accession="ENSP00000497305.1|ENST00000648836.1|ENSG00000255639.3|OTTHUMG00000168352.4|OTTHUMT00000500926.1|AC005833.1-201|AC005833.1|567" score="0.0" sequence="" >
    </ProteinHit>
    <UserParam type="string" name="protein_group_2714" value="1.0,PH_9008"/>
    <UserParam type="string" name="protein_group_111582" value="0.0,PH_186555"/>
</ProteinIdentification>
```

So I was thinking the reason that the peptide not being used for quantitation (and yielding that debugging message from ProteinQuatnifier) is that ProteinQuantifier takes both `ENSP00000266544.5` and `ENSP00000497305.1` as individual protein groups, while `ENSP00000497305.1` should be actually discarded because the ProteinProphet probability is zero. And then when I filter out the protein groups with probability of zero, it starts to give me reasonable results (that most proteins are quantified in most plexes). Apologize for not making it clear enough in the beginning and let me know if it makes sense!

**Comment by jpfeuffer** [comment]
Yes, that is correct. What I am saying is just, that this does not really affect the plex-completeness of proteins (or at least in a non-trivial way, since it depends on the completeness of the channels of the peptides that are shared).
I will look into creating indistinguishable group annotations during protXML conversion but I am really not sure if and where to put a filtering step (@timosachsenberg ?)

**Comment by zhuchcn** [comment]
But is it still true that the "leader" protein reported by ProteinProphet with probability of zero stops ProteinQuantifier to use its shared peptides to quantify its actual protein group (leader)? Adding the probability score to indistinguishable group during protXML conversion definitely is helpful! I think for us, filtering those protein groups of zero gives us results that we think are reasonable (and similar to the results in the original publication). But if this is an real issue, then it might affect other people from using the results correctly.

**Comment by timosachsenberg** [comment]
@jpfeuffer still relevant?

**Comment by jpfeuffer** [comment]
Yes

**Comment by timosachsenberg** [comment]
Fix submitted in #8960: the ProtXMLFile parser now skips probability=0 (subsumable) `<protein>` elements during parsing, preventing them from creating spurious indistinguishable groups that interfere with ProteinQuantifier. This matches how FragPipe/Philosopher and MaxQuant handle subsumable proteins.
